tooling/minimize/src/enums.rs

```diff
@@ -0,0 +1,121 @@
+use crate::*;
+mod rs {
+    pub use crate::rs::*;
+    pub use crate::rustc_target::abi::{Variants, FieldsShape, Primitive, TagEncoding, VariantIdx};
+}
+
+pub fn translate_enum<'tcx>(
+    ty: rs::Ty<'tcx>,
+    adt_def: rs::AdtDef<'tcx>,
+    sref: rs::GenericArgsRef<'tcx>,
+    tcx: rs::TyCtxt<'tcx>,
+) -> Type {
+    let a = rs::ParamEnv::empty().and(ty);
+    let layout = tcx.layout_of(a).unwrap().layout;
+    let size = translate_size(layout.size());
+    let align = translate_align(layout.align().abi);
+
+    let Type::Int(discriminant_ty) = translate_ty(ty.discriminant_ty(tcx), tcx) else {
+        panic!("Discriminant type is not integer!")
+    };
+
+    let (variants, discriminator) = match layout.variants() {
+        rs::Variants::Single { index } => {
+            let fields = translate_fields(layout.fields(), adt_def.variant(*index), sref, tcx);
+            ([(Int::ZERO, Variant { ty: Type::Tuple { fields, size, align }, tagger: Map::new() })].into_iter().collect(), Discriminator::Known(0.into()))
+        },
+        rs::Variants::Multiple {
+            tag,
+            tag_encoding,
+            tag_field,
+            variants,
+        } => {
+
+            // compute the offset of the tag for the tagger and discriminator construction
+            let tag_offset: Offset = translate_size(layout.fields().offset(*tag_field));
+            let tag_ty = match tag.primitive() {
+                rs::Primitive::Int(ity, signed) => IntType { signed: if signed { Signedness::Signed } else { Signedness::Unsigned }, size: translate_size(ity.size()) },
+                _ => panic!("enum tag has invalid primitive type"),
+            };
+
+            // translate the variants
+            let mut translated_variants = Map::new();
+            let mut discriminator_children = Map::new();
+            for (variant_idx, variant_def) in adt_def.variants().iter_enumerated() {
+                let fields = translate_fields(&variants[variant_idx].fields, &variant_def, sref, tcx);
+                let discr = adt_def.discriminant_for_variant(tcx, variant_idx);
+                let discr_int = int_from_bits(discr.val, discriminant_ty);
+                let (tagger, tag) = match tag_encoding {
+                    rs::TagEncoding::Direct => (
+                        // direct tagging places the discriminant in the tag for all variants
+                        [(tag_offset, (tag_ty, discr_int))].into_iter().collect(),
```

Comment:
Wow there's some magic type inference going on here.
```suggestion
                        [(tag_offset, (tag_ty, discr_int))].into_iter().collect::<Map<_, _>>(),
```